mypy/semanal.py

```diff
@@ -1894,10 +1894,14 @@ def process_imported_symbol(self,
                                 fullname: str,
                                 module_public: bool,
                                 context: ImportBase) -> None:
-        module_hidden = not module_public and not (
-            # `from package import module` should work regardless of whether package
-            # re-exports module
-            isinstance(node.node, MypyFile) and fullname in self.modules
+        module_hidden = not module_public and (
+            # `from package import submodule` should work regardless of whether package
+            # re-exports submodule, so we shouldn't hide it
+            not isinstance(node.node, MypyFile)
+            or fullname not in self.modules
+            # but given `from somewhere import random_unrelated_module` we should hide
+            # random_unrelated_module
+            or not fullname.startswith(self.cur_mod_id)
```

Comment:
What if there's a partial prefix, e.g. `foo.ba` vs `foo.bar`?

Reply:
Thanks, was a little sloppy. Fixed and added a test.